crates/anvil-polkadot/src/substrate_node/mining_engine.rs

```diff
@@ -0,0 +1,227 @@
+            let mode_changed = current_mode.as_ref().is_none_or(|m| *m != mode);
+
+            if mode_changed {
+                current_mode = Some(mode);
+                std::task::Poll::Ready(Some(()))
+            } else {
+                engine.inner.register(cx);
+                std::task::Poll::Pending
+            }
+        });
+        tokio::select! {
+            _ = rebuild_streams_future.next() => {
+                let mode = current_mode.unwrap_or(MiningMode::None);
+                interval_mining_stream = if let Some(tick) = match mode {
+                    MiningMode::Interval {tick} | MiningMode::MixedMining { tick } => Some(tick),
+                    _ => None,
+                } {
+                    let stream = unfold(interval(Duration::from_secs(tick)), |mut interval| async {
+                        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+                        interval.tick().await;
+                        Some((EngineCommand::SealNewBlock { create_empty: true, finalize: true, parent_hash: None, sender: None }, interval))
+                    }).fuse();
+                    Some(Box::pin(stream))
+                }
+                else {
+                    None
+                };
+                auto_mining_stream = if matches!(mode, MiningMode::AutoMining | MiningMode::MixedMining { .. }) {
+                    let stream = engine.transaction_pool.import_notification_stream().map(|_| EngineCommand::SealNewBlock { create_empty: false, finalize: true, parent_hash: None, sender: None })
+                        .fuse();
+                    Some(Box::pin(stream))
+                } else {
+                    None
+                };
+            },
+            // Poll the interval stream if it exists
+            maybe_cmd = async {
+                match interval_mining_stream.as_mut() {
+                    Some(stream) => stream.next().await,
+                    // return a future that never resolves
+                    None => futures::future::pending().await,
+                }
+            } => {
+                if let Some(seal_command) = maybe_cmd {
+                    info!("---->Interval miner ticked at: {}", Local::now().format("%Y-%m-%d %H:%M:%S"));
+                    if command_sink.send(seal_command).await.is_err() {
+                        break;
+                    }
+                }
+            },
+            // Poll the automining stream if it exists
+            maybe_cmd = async {
+                match auto_mining_stream.as_mut() {
+                    Some(stream) => stream.next().await,
+                    None => futures::future::pending().await,
+                }
+                } => {
+                    if let Some(seal_command) = maybe_cmd {
+                        if command_sink.send(seal_command).await.is_err(){
+                            break;
+                        }
+                    }
+                }
```

Comment:
Double check in Anvil what happens with the timer if a transactions arrives before the tick. Does it continue or is restarted?